dotnet/src/Experimental/Process.LocalRuntime/LocalProcess.cs

```diff
@@ -299,21 +336,44 @@ private void EnqueueStepMessages(LocalStep step, Queue<LocalMessage> messageChan
         var allStepEvents = step.GetAllEvents();
         foreach (var stepEvent in allStepEvents)
         {
-            foreach (var edge in step.GetEdgeForEvent(stepEvent.Id!))
+            // Emit the event out of the process (this one) if it's visibility is public.
+            if (stepEvent.Visibility == KernelProcessEventVisibility.Public)
             {
-                var target = edge.OutputTarget;
-                Dictionary<string, object?> parameterValue = new();
-                if (!string.IsNullOrWhiteSpace(target.ParameterName))
-                {
-                    parameterValue.Add(target.ParameterName!, stepEvent.Data);
-                }
+                base.EmitEvent(stepEvent);
+            }
 
-                LocalMessage newMessage = new(edge.SourceStepId, target.StepId, target.FunctionName, parameterValue);
-                messageChannel.Enqueue(newMessage);
+            // Get the edges for the event and queue up the messages to be sent to the next steps.
+            foreach (var edge in step.GetEdgeForEvent(stepEvent.Id!))
+            {
+                LocalMessage message = LocalMessageFactory.CreateFromEdge(edge, stepEvent.Data);
+                messageChannel.Enqueue(message);
             }
         }
     }
 
+    /// <summary>
+    /// Builds a <see cref="KernelProcess"/> from the current <see cref="LocalProcess"/>.
+    /// </summary>
+    /// <returns>An instance of <see cref="KernelProcess"/></returns>
+    /// <exception cref="InvalidOperationException"></exception>
+    private async Task<KernelProcess> ToKernelProcessAsync()
+    {
+        var processState = new KernelProcessState(this.Name, this.Id);
+        var stepTasks = this._steps.Select(step => step.ToKernelProcessStepInfoAsync()).ToList();
+        var steps = await Task.WhenAll(stepTasks).ConfigureAwait(false);
+        return new KernelProcess(processState, steps, this._outputEdges);
+    }
+
+    /// <summary>
+    /// When the process is used as a step within another process, this method will be called
+    /// rather than ToKernelProcessAsync when extracting the state.
+    /// </summary>
+    /// <returns>A <see cref="Task{T}"/> where T is <see cref="KernelProcess"/></returns>
+    internal override async Task<KernelProcessStepInfo> ToKernelProcessStepInfoAsync()
```

Comment:
`KernelProcess` inherits from `KernelProcessStepInfo` and this method is just auto-casting.  Since this is for `internal` use, why not just call `ToKernelProcessAsync` and save some wrapper code?